testutil/network.go

```diff
@@ -0,0 +1,335 @@
+type Config struct {
+	GenesisState    map[string]json.RawMessage
+	TimeoutCommit   time.Duration
+	ChainID         string
+	NumValidators   int
+	BondDenom       string
+	MinGasPrices    string
+	Passphrase      string
+	AccountTokens   sdk.Int
+	StakingTokens   sdk.Int
+	BondedTokens    sdk.Int
+	PruningStrategy string
+	EnableLogging   bool
+}
+
+// DefaultConfig returns a sane default configuration suitable for nearly all
+// testing requirements.
+func DefaultConfig() Config {
+	return Config{
+		GenesisState:    simapp.ModuleBasics.DefaultGenesis(cdc),
+		TimeoutCommit:   2 * time.Second,
+		ChainID:         "chain-" + tmrand.NewRand().Str(6),
+		NumValidators:   4,
+		BondDenom:       sdk.DefaultBondDenom,
+		MinGasPrices:    fmt.Sprintf("0.000006%s", sdk.DefaultBondDenom),
+		Passphrase:      clientkeys.DefaultKeyPass,
+		AccountTokens:   sdk.TokensFromConsensusPower(1000),
+		StakingTokens:   sdk.TokensFromConsensusPower(500),
+		BondedTokens:    sdk.TokensFromConsensusPower(100),
+		PruningStrategy: storetypes.PruningOptionNothing,
+	}
+}
+
+type (
+	// Network defines a local in-process testing network using SimApp. It can be
+	// configured to start any number of validators, each with its own RPC and API
+	// clients. Typically, this test network would be used in client and integration
+	// testing where user input is expected.
+	Network struct {
+		T          *testing.T
+		BaseDir    string
+		Validators []*Validator
+	}
+
+	// Validator defines an in-process Tendermint validator node. Through this object,
+	// a client can make RPC and API calls and interact with any client command
+	// or handler.
+	Validator struct {
+		AppConfig  *srvconfig.Config
+		ClientCtx  client.Context
+		Ctx        *server.Context
+		Dir        string
+		NodeID     string
+		PubKey     crypto.PubKey
+		Moniker    string
+		RPCAddress string
+		P2PAddress string
+		Address    sdk.AccAddress
+		ValAddress sdk.ValAddress
+		RPCClient  tmclient.Client
+
+		tmNode *node.Node
+		api    *api.Server
+	}
+)
+
+func NewTestNetwork(t *testing.T, cfg Config) *Network {
+	baseDir, err := ioutil.TempDir(os.TempDir(), cfg.ChainID)
+	require.NoError(t, err)
+
+	network := &Network{
+		T:          t,
+		BaseDir:    baseDir,
+		Validators: make([]*Validator, cfg.NumValidators),
+	}
+
+	t.Log("preparing test network...")
+
+	monikers := make([]string, cfg.NumValidators)
+	nodeIDs := make([]string, cfg.NumValidators)
+	valPubKeys := make([]crypto.PubKey, cfg.NumValidators)
+
+	var (
+		genAccounts []authtypes.GenesisAccount
+		genBalances []banktypes.Balance
+		genFiles    []string
+	)
+
+	buf := bufio.NewReader(os.Stdin)
+
+	// generate private keys, node IDs, and initial transactions
+	for i := 0; i < cfg.NumValidators; i++ {
+		appCfg := srvconfig.DefaultConfig()
+		appCfg.Pruning = cfg.PruningStrategy
+		appCfg.MinGasPrices = cfg.MinGasPrices
+		appCfg.API.Enable = true
+		appCfg.API.Swagger = false
+		appCfg.Telemetry.Enabled = false
```

Comment:
May be better to allow different validtor configs?

Reply:
WDYM different validator configs? There's only a single configuration passed to `NewTestNetwork`.

Reply:
May be not needed for integration tests but I was thinking if `NetTestNetwork()` takes an array of configs (validator configs like Pruning, MinGasPrices, etc) instead of generating it here

Reply:
Good idea. I think we can explore that later down the line, as it dramatically increases the complexity of bootstrapping.

The real neat thing is it's super simple to start a test network for 99% of use-cases:

```go
n := NewTestNetwork(s.T(), DefaultConfig())
defer n.Clean()
```

If, however, as we're refactoring and implementing new integration tests, it becomes apparent we need such functionality, we can _extend_ (add a new function) that allows for explicit validator configurations. WDYT?

Reply:
That sounds good